test/features/app/spec.js

```diff
@@ -186,77 +163,132 @@ function cmp(result, expected, errMsg) {
+            if (val.constructor === Object) {
+                validateBody(resBody[key], val);
+            } else if (val.constructor === Array) {
+                validateArray(val, resBody[key], key);
+            } else {
+                cmp(resBody[key], val, `${key} body field mismatch!`);
+            }
+        });
+    } else if (Array.isArray(expBody)) {
+        validateArray(expBody, resBody, 'body');
+    } else {
+        cmp(resBody, expBody, 'Body mismatch!');
     }
+    return true;
+}
 
-    // compare the bodies
-    cmp(res.body, expRes.body, 'Body mismatch!');
+function validateTestResponse(testCase, res) {
+    const expRes = testCase.response;
 
-    return true;
+    assert.deepEqual(res.status, expRes.status);
 
-}
+    if (expRes.headers && !res.headers) {return false; } // eslint-disable-line
 
+    Object.keys(expRes.headers).forEach((key) => {
+        const val = expRes.headers[key];
+        // eslint-disable-next-line
+        assert.deepEqual(res.headers.hasOwnProperty(key), true, `Header ${key} not found in response!`);
+        cmp(res.headers[key], val, `${key} header mismatch!`);
+    });
 
-describe('Swagger spec', function() {
+    validateBody(res.body || '', expRes.body);
+}
 
-    // the variable holding the spec
-    let spec = staticSpecLoad();
-    // default params, if given
-    let defParams = spec['x-default-params'] || {};
+describe('Swagger spec', function () {
 
-    this.timeout(20000); // eslint-disable-line no-invalid-this
+    this.timeout(20000);
 
-    before(() => {
-        return server.start();
-    });
+    before(() => server.start());
+    after(() => server.stop());
 
     it('get the spec', () => {
-        return preq.get(`${server.config.uri}?spec`)
+        baseUrl = server.config.uri;
+        return preq.get(`${baseUrl}?spec`)
         .then((res) => {
             assert.status(200);
             assert.contentType(res, 'application/json');
             assert.notDeepEqual(res.body, undefined, 'No body received!');
+            // save a copy
             spec = res.body;
+            return spec;
+        })
+        .then((spec) => {
+            const routeTests = () => {
+                before(() => server.start());
+                after((done) => setTimeout(() => {
+                    server.stop()
+                    .then(() => done());
+                }, 3000));
+
+                constructTests(spec).forEach((testCase) => {
+                    it(testCase.title, function (done) {
+                        return preq(testCase.request)
+                        .then((res) => {
+                            assert.status(res, testCase.response.status);
+                            validateTestResponse(testCase, res);
+                        }, (err) => {
+                            assert.status(err, testCase.response.status);
+                            validateTestResponse(testCase, err);
+                        })
+                        .then(done());
```

Comment:
This is the root of all your problems. 

There are several styles of writing async tests with mocha: using callbacks (and the `done` function) or using promises and returning them from a test. The 2 must not be mixed unless you know exactly what you're doing (even in that case it's not recommended). Here you're mixing incorrect usage of `done` and a promise return, which makes mocha get crazy.

The tests are reported to fail incorrectly because of this `done` as well - in case of a failure you need to provide an error object to `done` and you are never doing it - of course the tests will be reported as passing no matter what. If you added a `catch` clause with a call to `done` as well, the tests would've failed. 

Why did you add a `done` in the first place? we are already returning a promise, that's enough.

Reply:
The tests would hang without it in a previous iteration of the changes. I removed it